### PR TITLE
MGMT-9821: Add Conditions to AgentMachines

### PR DESCRIPTION
**Pull request by avishayt**
Create conditions for finding an agent and installing the node, and
implement the Ready Condition which is bubbled up to the nodepool.

Status when no agents available:
```
    conditions:
    - lastTransitionTime: "2022-03-31T10:17:31Z"
      message: 0 of 5 completed
      reason: NoSuitableAgentFound
      severity: Warning
      status: "False"
      type: Ready
    - lastTransitionTime: "2022-03-31T10:17:31Z"
      message: Agent not yet reserved
      reason: AgentNotYetFound
      severity: Info
      status: "False"
      type: AgentRequirementsMet
    - lastTransitionTime: "2022-03-31T10:17:31Z"
      reason: NoSuitableAgentFound
      severity: Warning
      status: "False"
      type: AgentReserved
    - lastTransitionTime: "2022-03-31T10:17:31Z"
      message: Agent not yet reserved
      reason: AgentNotYetFound
      severity: Info
      status: "False"
      type: AgentSpecSynced
    - lastTransitionTime: "2022-03-31T10:17:31Z"
      message: Agent not yet reserved
      reason: AgentNotYetFound
      severity: Info
      status: "False"
      type: AgentValidated
    - lastTransitionTime: "2022-03-31T10:17:31Z"
      message: Agent not yet reserved
      reason: AgentNotYetFound
      severity: Info
      status: "False"
      type: Installed
    ready: false
```

If validations are failing:
```
    conditions:
    - lastTransitionTime: "2022-03-31T10:06:50Z"
      message: 2 of 5 completed
      reason: ValidationsFailing
      severity: Error
      status: "False"
      type: Ready
    - lastTransitionTime: "2022-03-31T10:06:50Z"
      message: The agent is not ready to begin the installation
      reason: AgentNotReady
      severity: Error
      status: "False"
      type: AgentRequirementsMet
    - lastTransitionTime: "2022-03-31T10:06:20Z"
      status: "True"
      type: AgentReserved
    - lastTransitionTime: "2022-03-31T10:06:20Z"
      status: "True"
      type: AgentSpecSynced
    - lastTransitionTime: "2022-03-31T10:06:50Z"
      message: 'The agent''s validations are failing: Missing role assignment,Ignition
        is not downloadable,Host couldn''t synchronize with any NTP server'
      reason: ValidationsFailing
      severity: Error
      status: "False"
      type: AgentValidated
    - lastTransitionTime: "2022-03-31T10:06:50Z"
      message: The installation has not yet started
      reason: InstallationNotStarted
      severity: Info
      status: "False"
      type: Installed
    ready: false
```

Starting installation:
```
    conditions:
    - lastTransitionTime: "2022-03-31T10:19:41Z"
      message: 4 of 5 completed
      reason: InstallationNotStarted
      severity: Info
      status: "False"
      type: Ready
    - lastTransitionTime: "2022-03-31T10:19:41Z"
      status: "True"
      type: AgentRequirementsMet
    - lastTransitionTime: "2022-03-31T10:19:41Z"
      status: "True"
      type: AgentReserved
    - lastTransitionTime: "2022-03-31T10:19:41Z"
      status: "True"
      type: AgentSpecSynced
    - lastTransitionTime: "2022-03-31T10:19:41Z"
      status: "True"
      type: AgentValidated
    - lastTransitionTime: "2022-03-31T10:19:41Z"
      message: The installation has not yet started
      reason: InstallationNotStarted
      severity: Info
      status: "False"
      type: Installed
    ready: false
```

### CODE REVIEW

File: controllers/agentmachine_controller.go [hunk 1]
```diff
@@ -124,20 +126,20 @@ func (r *AgentMachineReconciler) Reconcile(ctx context.Context, req ctrl.Request
 	if agentMachine.Status.AgentRef == nil {
 		foundAgent, result, err := r.findAgent(ctx, log, agentMachine, agentCluster.Status.ClusterDeploymentRef, machineConfigPool, ignitionTokenSecretRef)
 		if err != nil || foundAgent == nil {
-			return result, err
+			return r.updateStatus(ctx, log, agentMachine, result, err, err == nil)
```
Comment (eranco74) [inline controllers/agentmachine_controller.go:129]: I think this is clearer 
```suggestion
			return r.updateStatus(ctx, log, agentMachine, result, err, foundAgent == nil)
```

File: controllers/agentmachine_controller.go [hunk 2]
```diff
@@ -425,40 +423,61 @@ func isValidAgent(agent *aiv1beta1.Agent, agentMachine *capiproviderv1alpha1.Age
...
-				log.Info("Updating agentMachine status to Ready=true")
-				agentMachine.Status.Ready = true
-			} else if condition.Status == "False" {
-				if condition.Reason == aiv1beta1.InstallationFailedReason {
-					agentMachine.Status.FailureReason = (*clustererrors.MachineStatusError)(&condition.Reason)
-					message := condition.Message
-					agentMachine.Status.FailureMessage = &message
-				}
+		agent := &aiv1beta1.Agent{}
+		agentRef := types.NamespacedName{Name: agentMachine.Status.AgentRef.Name, Namespace: agentMachine.Status.AgentRef.Namespace}
```
Comment (eranco74) [inline controllers/agentmachine_controller.go:439]: Consider extracting it to another method

File: controllers/agentmachine_controller.go [hunk 3]
```diff
@@ -425,40 +423,61 @@ func isValidAgent(agent *aiv1beta1.Agent, agentMachine *capiproviderv1alpha1.Age
...
+		if err := r.AgentClient.Get(ctx, agentRef, agent); err != nil {
+			log.WithError(err).Errorf("Failed to get agent %s", agentRef)
+			return ctrl.Result{RequeueAfter: defaultRequeueAfterOnError}, err
+		}
+
+		agentInstalledCondition := openshiftconditionsv1.FindStatusCondition(agent.Status.Conditions, aiv1beta1.InstalledCondition)
```
Comment (eranco74) [inline controllers/agentmachine_controller.go:445]: Same

File: controllers/agentmachine_controller.go [hunk 4]
```diff
@@ -425,40 +423,61 @@ func isValidAgent(agent *aiv1beta1.Agent, agentMachine *capiproviderv1alpha1.Age
...
-		return ctrl.Result{Requeue: true}, nil
+		return ctrl.Result{RequeueAfter: defaultRequeueAfterOnError}, nil
 	}
 	if agentMachine.Status.Ready {
 		// No need to requeue in case the agentMachine is ready
 		return ctrl.Result{}, nil
 	}
-	return ctrl.Result{RequeueAfter: defaultRequeueWaitingForAgentToBeInstalled}, nil
+	if installationInProgress && origErr == nil {
+		return ctrl.Result{RequeueAfter: defaultRequeueWaitingForAgentToBeInstalled}, nil
```
Comment (enxebre) [inline controllers/agentmachine_controller.go:477]: I'm curious why do you need to requeueAfter? Isn't this watching the agent events that would trigger a reconcile when it's intalled?

File: controllers/agentmachine_controller.go [hunk 5]
```diff
@@ -124,20 +126,21 @@ func (r *AgentMachineReconciler) Reconcile(ctx context.Context, req ctrl.Request
 	if agentMachine.Status.AgentRef == nil {
 		foundAgent, result, err := r.findAgent(ctx, log, agentMachine, agentCluster.Status.ClusterDeploymentRef, machineConfigPool, ignitionTokenSecretRef)
 		if err != nil || foundAgent == nil {
-			return result, err
+			// Update the status, err == nil indicates that we didn't find an agent because one didn't exist, not due to a retryable error
```
Comment (eranco74) [inline controllers/agentmachine_controller.go:129]: The comment is irrelevant

**Pull request merged**